### PR TITLE
Pods 2177

### PR DESCRIPTION
**Pull request by versatil56**

### CODE REVIEW

File: app/connectors/DesConnector.scala [hunk 1]
```diff
@@ -163,6 +165,13 @@ class DesConnectorImpl @Inject()(
   }
 
   private def validateJson(json: JsValue): PsaSubscription ={
+
+    if (json.transform(psaSubscriptionDetailsTransformer.transformToUserAnswers).isSuccess)
+      Logger.warn("PensionAdministratorSuccessfulMapToUserAnswers")
```
Comment (grantrobson) [inline app/connectors/DesConnector.scala]: Should a successful response yield a WARN logging message? Info or debug would be better don't you think? Warn implies there's a potential problem.

File: app/utils/JsonTransformations/AddressTransformer.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      ((userAnswersPath \ 'addressLine4).json.copyFrom((desAddressPath \ 'line4).json.pick)
+        orElse doNothing) and
+      ((userAnswersPath \ 'postalCode).json.copyFrom((desAddressPath \ 'postalCode).json.pick)
+        orElse doNothing) and
+      (userAnswersPath \ 'countryCode).json.copyFrom((desAddressPath \ 'countryCode).json.pick) reduce
+  }
+
+  def getDifferentAddress(userAnswersPath: JsPath, desAddressPath: JsPath): Reads[JsObject] = {
+    (userAnswersPath \ 'addressLine1).json.copyFrom((desAddressPath \ 'line1).json.pick) and
+      (userAnswersPath \ 'addressLine2).json.copyFrom((desAddressPath \ 'line2).json.pick) and
```
Comment (grantrobson) [inline app/utils/JsonTransformations/AddressTransformer.scala]: As most of these address lines are repeated I'd separate the duplicate code into a private def.

File: app/utils/JsonTransformations/AddressTransformer.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  def getAddress(userAnswersPath: JsPath, desAddressPath: JsPath): Reads[JsObject] = {
+    (userAnswersPath \ 'addressLine1).json.copyFrom((desAddressPath \ 'line1).json.pick) and
+      (userAnswersPath \ 'addressLine2).json.copyFrom((desAddressPath \ 'line2).json.pick) and
+      ((userAnswersPath \ 'addressLine3).json.copyFrom((desAddressPath \ 'line3).json.pick)
+        orElse doNothing) and
+      ((userAnswersPath \ 'addressLine4).json.copyFrom((desAddressPath \ 'line4).json.pick)
+        orElse doNothing) and
+      ((userAnswersPath \ 'postalCode).json.copyFrom((desAddressPath \ 'postalCode).json.pick)
+        orElse doNothing) and
+      (userAnswersPath \ 'countryCode).json.copyFrom((desAddressPath \ 'countryCode).json.pick) reduce
```
Comment (grantrobson) [inline app/utils/JsonTransformations/AddressTransformer.scala]: This gives a compile warning - get rid of by surrounding with brackets and saying .reduce

File: app/utils/JsonTransformations/AddressTransformer.scala [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  def getPreviousAddress(path: JsPath) : Reads[JsObject] = {
+    (__ \ 'previousAddressDetails \ 'previousAddress).read[JsObject].flatMap { _ =>
+      getDifferentAddress(path , __ \ 'previousAddressDetails \ 'previousAddress)
+    } orElse doNothing
+  }
+
+  val getAddressYearsBasedOnLegalStatus: Reads[JsObject] = {
+    legalStatusTransformer.returnPathBasedOnLegalStatus(__ \ 'individualAddressYears, __ \ 'companyAddressYears, __ \ 'partnershipAddressYears).flatMap { addressYearsPath =>
+      getAddressYears(__ \ "psaSubscriptionDetails", addressYearsPath)
```
Comment (grantrobson) [inline app/utils/JsonTransformations/AddressTransformer.scala]: Lines too long around here

File: app/utils/JsonTransformations/AdviserTransformer.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    ((__ \ 'adviserDetails \ 'name).json.copyFrom((__ \ 'psaSubscriptionDetails \ 'declarationDetails \ 'pensionAdvisorDetails \ 'name).json.pick)
+      orElse doNothing) and
+      ((__ \ 'adviserDetails \ 'email).json.copyFrom(
+        (__ \ 'psaSubscriptionDetails \ 'declarationDetails \ 'pensionAdvisorDetails \ 'contactDetails \ 'email).json.pick)
+        orElse doNothing) and
+      ((__ \ 'adviserDetails \ 'phone).json.copyFrom(
+        (__ \ 'psaSubscriptionDetails \ 'declarationDetails \ 'pensionAdvisorDetails \ 'contactDetails \ 'telephone).json.pick)
+        orElse doNothing) and
+      (addressTransformer.getDifferentAddress(__ \ 'adviserAddress, __ \ 'psaSubscriptionDetails \ 'declarationDetails \ 'pensionAdvisorDetails \ 'addressDetails)
+        orElse doNothing) reduce
```
Comment (grantrobson) [inline app/utils/JsonTransformations/AdviserTransformer.scala]: Line too long plus see previous comment on reduce method

File: app/utils/JsonTransformations/DirectorOrPartnerTransformer.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        (__ \ s"$path$ninoOrUtr" \ s"has$ninoOrUtr").json.put(JsBoolean(false)) reduce
+    }
+  }
+
+  private def getDirectorOrPartnerContactDetails(directorOrPartner: String): Reads[JsObject] = {
+    (__ \ s"${directorOrPartner}ContactDetails" \ 'phone).json.copyFrom((__ \ 'correspondenceCommonDetails \ 'contactDetails \ 'telephone).json.pick) and
+      (__ \ s"${directorOrPartner}ContactDetails" \ 'email).json.copyFrom((__ \ 'correspondenceCommonDetails \ 'contactDetails \ 'email).json.pick) reduce
+  }
+
+  def getDirectorOrPartner(directorOrPartner: String): Reads[JsObject] = (__ \ s"${directorOrPartner}Details" \ 'firstName).json.copyFrom((__ \ 'firstName).json.pick) and
```
Comment (grantrobson) [inline app/utils/JsonTransformations/DirectorOrPartnerTransformer.scala]: A few of lines too long here plus see previous reduce comment

File: app/connectors/DesConnector.scala (hunk 1 above)
Comment (versatil56) [reply app/connectors/DesConnector.scala]: changed to info as suggested

File: app/utils/JsonTransformations/AddressTransformer.scala (hunk 2 above)
Comment (versatil56) [reply app/utils/JsonTransformations/AddressTransformer.scala]: created common private method as requested

File: app/utils/JsonTransformations/AddressTransformer.scala (hunk 4 above)
Comment (versatil56) [reply app/utils/JsonTransformations/AddressTransformer.scala]: made shorter

**Pull request merged**